File: pfm-sdk/src/main/java/com/tink/pfmsdk/ClientDataStorage.java
package com.tink.pfmsdk;

import android.content.Context;
import android.content.SharedPreferences;
import java.util.UUID;
import se.tink.privacy.Clearable;
import se.tink.privacy.DataWipeManager;

public class ClientDataStorage implements Clearable {

	private static ClientDataStorage instance;

	private static final String PREFS_NAME = "client-data-storage";

	private final SharedPreferences settings;
	private SharedPreferences.Editor editor;
	private String uuid;

	public static synchronized ClientDataStorage sharedInstance(Context context) {
		if (instance == null) {
			instance = new ClientDataStorage(context);
			DataWipeManager.sharedInstance().register(instance);
		}
		return instance;
	}

	private ClientDataStorage(Context context) {
		settings = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
		editor = settings.edit();
	}

	public String getUUID() {
		String uuid = settings.getString("tinkDeviceId", null);
		if (uuid == null) {
			uuid = this.uuid;
			setUUID(uuid);
		}
		if (uuid == null) {
			uuid = UUID.randomUUID().toString();
			setUUID(uuid);
		}
		return uuid;
	}

	public void setUUID(String uuid) {
		this.uuid = uuid;
		editor.putString("tinkDeviceId", uuid);
		editor.apply();
	}

	public void setLastVisitedPageInOverview(int page) {
		editor.putInt("lastViewPageInOverview", page);
		editor.apply();
	}

	public int getLastVisitedPageInOverview() {
		return settings.getInt("lastViewPageInOverview", 1); // left to spend is default
	}

	public void clear() {
		editor.clear().apply();
	}

}

File: pfm-ui/src/main/java/com/tink/pfmui/ClientDataStorage.kt
package com.tink.pfmui

import android.content.Context
import android.content.SharedPreferences
import se.tink.privacy.Clearable
import se.tink.privacy.DataWipeManager
import java.util.UUID

internal class ClientDataStorage private constructor(context: Context) : Clearable {
    private val settings: SharedPreferences
    private val editor: SharedPreferences.Editor
    var uuid: String? = null
    var uUID: String?
        get() {
            var uuid = settings.getString(DEVICE_ID, null)
            if (uuid == null) {
                uuid = this.uuid
                uUID = uuid
            }
            if (uuid == null) {
                uuid = UUID.randomUUID().toString()
                uUID = uuid
            }
            return uuid
        }
        set(uuid) {
            this.uuid = uuid
            editor.putString(DEVICE_ID, uuid)
            editor.apply()
        }

    // left to spend is default
    var lastVisitedPageInOverview: Int
        get() = settings.getInt(LAST_VIEW_PAGE_IN_OVERVIEW, 1) // left to spend is default
        set(page) {
            editor.putInt(LAST_VIEW_PAGE_IN_OVERVIEW, page)
            editor.apply()
        }

    override fun clear() {
        editor.clear().apply()
    }

    companion object {
        private const val PREFS_NAME = "client-data-storage"
        private const val DEVICE_ID = "tinkDeviceId"
        private const val LAST_VIEW_PAGE_IN_OVERVIEW = "lastViewPageInOverview"

        private var instance: ClientDataStorage? = null
        @JvmStatic
        @Synchronized
        fun sharedInstance(context: Context): ClientDataStorage =
            instance
                ?.let { it }
                ?: ClientDataStorage(context)
                    .let {
                        DataWipeManager.sharedInstance().register(it)
                        instance = it
                        it
                    }

    }

    init {
        settings = context.getSharedPreferences(
            PREFS_NAME,
            Context.MODE_PRIVATE
        )
        editor = settings.edit()
    }
}